src/hypothesis/strategies.py

```diff
@@ -910,6 +912,110 @@ def do_draw(self, data):
     return PermutationStrategy()
 
 
+@defines_strategy
+def datetimes(min_datetime=dt.datetime.min, max_datetime=dt.datetime.max,
+              timezones=None):
+    """A strategy for generating datetimes, which may be timezone-aware.
+
+    This strategy works by drawing a naive datetime between ``min_datetime``
+    and ``max_datetime``, which must both be naive (have no timezone).
+    If ``timezones`` is None, this value is returned.
+
+    If ``timezones`` is not None, it must be a strategy that generates tzinfo
+    objects (or None, which is valid for naive datetimes).  A value drawn
+    from this strategy will be added to a naive datetime, and the resulting
+    tz-aware datetime returned.
+
+    .. note::
+        tz-aware datetimes from this strategy may be ambiguous or non-existent
+        due to daylight savings, leap seconds, timezone and calendar
+        adjustments, etc.  This is intentional, as malformed timestamps are a
+        common source of bugs.  Consider validating all datetime values, or
+        working purely in UTC to avoid such issues.
+
+    There are a number of ways you can create a timezones strategy, depending
+    mostly on your timezone objects (e.g. from the standard library,
+    ``datetutil``, or ``pytz``).  In each case, you should probably use
+    :py:func:`sampled_from` to sample from a sequence.  To minimise
+    effectively, consider composing strategies for None, UTC, and other
+    timezones in that order (e.g. using :py:func:`one_of`).
+
+    :py:func:`hypothesis.extra.timezones` requires the ``pytz`` package, but
+    provides all timezones in the Olsen database.  If you also want to allow
+    naive datetimes, combine strategies like ``none() | timezones()``.
+
+    """
+    from hypothesis.searchstrategy.datetime import DatetimeStrategy
+
+    check_type(dt.datetime, min_datetime, 'min_datetime')
+    check_type(dt.datetime, max_datetime, 'max_datetime')
+    if min_datetime.tzinfo is not None:
+        raise InvalidArgument('min_datetime=%r must not have tzinfo'
+                              % min_datetime)
+    if max_datetime.tzinfo is not None:
+        raise InvalidArgument('max_datetime=%r must not have tzinfo'
+                              % max_datetime)
+    check_valid_interval(min_datetime, max_datetime,
+                         'min_datetime', 'max_datetime')
+    if timezones is None:
+        if min_datetime == max_datetime:
+            return just(min_datetime)
+        timezones = none()
+    if not isinstance(timezones, SearchStrategy):
+        raise InvalidArgument(
+            'timezones=%r must be a SearchStrategy that can provide tzinfo '
+            'for datetimes (either None or dt.tzinfo objects)' % (timezones,))
```

Comment:
I was thinking that one obvious mistake would be to supply e.g. `timezones=[None, pytz.UTC]` instead of a strategy.  Above, we know that it's either None or a datetime so there's no need.